qiskit/transpiler/passes/scheduling/padding.py

```diff
@@ -205,11 +232,307 @@ def _pad(
+        else:
+            if sum(self._spacing) != 1 or any(a < 0 for a in self._spacing):
+                raise TranspilerError(
+                    "The spacings must be given in terms of fractions "
+                    "of the slack period and sum to 1."
+                )
+
+        # Check if DD sequence is identity
+        if num_pulses != 1:
+            if num_pulses % 2 != 0:
+                raise TranspilerError("DD sequence must contain an even number of gates (or 1).")
+            noop = np.eye(2)
+            for gate in self._dd_sequence:
+                noop = noop.dot(gate.to_matrix())
+            if not matrix_equal(noop, IGate().to_matrix(), ignore_phase=True):
+                raise TranspilerError("The DD sequence does not make an identity operation.")
+            self._sequence_phase = np.angle(noop[0][0])
+
+        # Precompute qubit-wise DD sequence length for performance
+        for qubit in dag.qubits:
+            physical_index = dag.qubits.index(qubit)
+            if self._qubits and physical_index not in self._qubits:
+                continue
+
+            gate_length_sum = 0
+            for gate in self._dd_sequence:
+                try:
+                    # Check calibration.
+                    gate_length = dag.calibrations[gate.name][(physical_index, gate.params)]
+                    if gate_length % self._alignment != 0:
+                        # This is necessary to implement lightweight scheduling logic for this pass.
+                        # Usually the pulse alignment constraint and pulse data chunk size take
+                        # the same value, however, we can intentionally violate this pattern
+                        # at the gate level. For example, we can create a schedule consisting of
+                        # a pi-pulse of 32 dt followed by a post buffer, i.e. delay, of 4 dt
+                        # on the device with 16 dt constraint. Note that the pi-pulse length
+                        # is multiple of 16 dt but the gate length of 36 is not multiple of it.
+                        # Such pulse gate should be excluded.
+                        raise TranspilerError(
+                            f"Pulse gate {gate.name} with length non-multiple of {self._alignment} "
+                            f"is not acceptable in {self.__class__.__name__} pass."
+                        )
+                except KeyError:
+                    gate_length = self._durations.get(gate, physical_index)
+                gate_length_sum += gate_length
+                # Update gate duration. This is necessary for current timeline drawer, i.e. scheduled.
+                gate.duration = gate_length
+            self._dd_sequence_lengths[qubit] = gate_length_sum
+
+    def _pad(
+        self,
+        dag: DAGCircuit,
+        qubit: Qubit,
+        t_start: int,
+        t_end: int,
+        next_node: DAGNode,
+        prev_node: DAGNode,
+    ):
```

Comment:
i missed this in my review of the previous PR, but I just realized that _pad() replaces the run() method in these passes. Since this is the main method, i don't think it should be underscored (private). 
But also why not just use the `run` method? I think all other passes have a run method, and it is treated somewhat especially by the transpiler. For example in BasePass the `__call__` method has been defined to call `run()`. This `_pad()`  method would be inconsistent with that.

Reply:
No, this is a sub routine of `run` method to define a sequence to be interleaved. The `run` is the outer-loop of the `_pad` method and generation of new DAG circuit (i.e. management of extra data such as metadata, cal, etc...) and construction of non-idle time instruction are performed there. Thus subclass can remove the overhead of implementing the logic to parse `node_start_time` and compute idle time.

Reply:
ok i understand now. it's nice.